src/main/java/org/carlspring/cloud/storage/s3fs/S3Path.java

```diff
@@ -735,11 +736,20 @@ public int hashCode()
      */
     private String encode(String uri)
     {
-        // remove special case URI starting with //
-        uri = uri.replace("//", "/");
-        uri = uri.replaceAll(" ", "%20");
-
-        return uri;
+        try
+        {
+            // URL encode all characters, but then convert known allowed characters back
+            return URLEncoder.encode(uri, "UTF-8")
+                    .replace("%3A", ":")
+                    .replace("%2F", "/")
+                    .replace("+", "%20");
+        }
+        catch (UnsupportedEncodingException e)
+        {
+            // This should never happen unless there is something
+            // fundamentally broken with the running JVM.
+            throw new RuntimeException(e);
```

Comment:
We should definitely add a short message here and maybe change the type to Error?
A message on the other hand should definitely say something along the Line "URI could not be encoded into an URL. JVM is broken". Not sure i agree with the assesment though, but the URI and URL Standard are broken anyway.

Reply:
Yeah, this is code that is actually copied from Amazon's SDK, from what I saw.

Reply:
However you do it I think you should be very explicit about the behavior in the docs. Valid S3 Paths are not always going to be fully compatible with file paths encoded as URIs. You could follow the lead of FSx for Lustre which, when backed by S3, effectively hides some valid S3 paths which cannot be converted into POSIX file paths.